backend/repository/validator.go

```diff
@@ -0,0 +1,378 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package repository
+
+import (
+	"fmt"
+	"path/filepath"
+	"reflect"
+	"strings"
+
+	"github.com/pkg/errors"
+	"go.uber.org/thriftrw/compile"
+)
+
+// Validator validates a request in gateway repo, a validator
+// executes a group of rules
+type Validator interface {
+	Validate(*Repository, *UpdateRequest) error
+}
+
+// Rule validates client/endpoint config against compiled thrift modules
+type Rule interface {
+	validate([]ClientConfig, []EndpointConfig, map[string]*compile.Module) error
+}
+
+// SyntaxValidator implements validator and validates thrift syntax for codegen
+type SyntaxValidator struct {
+	rules []Rule
+}
+
+// NewSyntaxValidator returns a new Validator
+func NewSyntaxValidator() Validator {
+	rules := []Rule{
+		&RuleProxyConsistency{},
+	}
+	return &SyntaxValidator{
+		rules: rules,
+	}
+}
+
+// Validate prepare configs / modules for rule validations
+func (sv *SyntaxValidator) Validate(repo *Repository, req *UpdateRequest) error {
+	var (
+		clientCfgs    = req.ClientUpdates
+		endpointCfgs  = req.EndpointUpdates
+		thriftModules = make(map[string]*compile.Module)
+	)
+	for _, clientCfg := range clientCfgs {
+		if len(clientCfg.ThriftFile) > 0 {
+			thriftModules[clientCfg.ThriftFile] = nil
+		}
+	}
+	for _, endpointCfg := range endpointCfgs {
+		if len(endpointCfg.ThriftFile) > 0 {
+			thriftModules[endpointCfg.ThriftFile] = nil
+		}
+	}
+	cfg, err := repo.GatewayConfig()
+	if err != nil {
+		return nil
+	}
+	for thriftFile := range thriftModules {
+		rootPath := filepath.Join(repo.LocalDir(), cfg.ThriftRootDir, thriftFile)
+		tfm, err := compile.Compile(rootPath)
+		if err != nil {
+			return err
+		}
+		thriftModules[thriftFile] = tfm
+	}
+	for _, rule := range sv.rules {
+		if err := rule.validate(clientCfgs, endpointCfgs, thriftModules); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RuleProxyConsistency validates client / endpoint proxy consistency
+type RuleProxyConsistency struct{}
+
+func (r *RuleProxyConsistency) validate(clientCfgs []ClientConfig, endpointCfgs []EndpointConfig, thriftModules map[string]*compile.Module) error {
+	// trigger
+	if len(clientCfgs) != 1 || len(endpointCfgs) != 1 {
+		return nil
+	}
+	clientCfg := clientCfgs[0]
+	endpointCfg := endpointCfgs[0]
+	// check len(thriftFile) > 0
+	clientModule := thriftModules[clientCfg.ThriftFile]
+	endpointModule := thriftModules[endpointCfg.ThriftFile]
+	// endpoint method should be exposed from client
+	if exposedMethod, ok := clientCfg.ExposedMethods[endpointCfg.ClientMethod]; !ok || exposedMethod != endpointCfg.ThriftMethodName {
+		return NewRequestError(
+			ClientsExposedMethods, errors.New("methods not exposed in proxy client"))
+	}
+	funcPath := strings.Split(endpointCfg.ThriftMethodName, "::")
+	clientFunc, err := ResolveNode(funcPath, clientModule)
```

Comment:
This does not seem correct. it should respect `endpointCfg.ClientMethod` not `endpointCfg.ThriftMethodName`.

The `ThriftMethodName` is for `endpointModule`; the `clientCfg.ExposedMethod[endpointCfg.ClientMethod]` should be used for `clientModule`